vtex-apps/b2b-organizations
# Impersonate Metrics

#### What problem is this solving?

Send events to analytics when a user impersonates/stop impersonating some other user, in order to understand better this behavior.

#### How to test it?
How to https://takeda--b2bstore005.myvtex.com/account#/organization and impersonate some user. Then, stop impersonating him/her. Events should appear in Analytics Redshift after one hour or so into "vtex"."schemaless"."b2b_suite_buyerorg_data_raw" schema.

To impersonate, in a user row, click on the dashed icon and chose personate user.

![image](https://github.com/vtex-apps/b2b-organizations/assets/1576737/7cdcb6ee-2bad-4214-9d92-dd14709e236d)

To stop personate, click the red button on top

![image](https://github.com/vtex-apps/b2b-organizations/assets/1576737/8f7002c6-4bac-4262-84b3-6bb800ed9dec)

[Workspace](https://takeda--b2bstore005.myvtex.com/account#/organization)


Types sent to analytics:
```
type ImpersonatePerson = {
  email: string
  buy_org_id: string
  cost_center_id: string
  cost_center_name: string
}

type ImpersonateUser = ImpersonatePerson
type ImpersonateTarget = ImpersonatePerson & { buy_org_name: string }

type ImpersonateFieldsMetric = {
  user: ImpersonateUser
  target: ImpersonateTarget
  date: string
}

type ImpersonateMetric = Metric & { fields: ImpersonateFieldsMetric }


type ImpersonateMetric = {
  kind: 'impersonate-ui-event'
  description: 'Impersonate User Action - UI'
}

type StopImpersonateMetric = {
  kind: 'stop-impersonate-ui-event'
  description: 'Stop Impersonate User Action - UI'
}

export type Metric = {
  name: 'b2b-suite-buyerorg-data'
  account: string
} & (ImpersonateMetric | StopImpersonateMetric)
```

Example of query and results:

```
SELECT
    payload.account,
    payload.kind,
    payload.fields.date,
    payload.fields.user.email as user_email,
    payload.fields.user.buy_org_id as user_buy_org,
    payload.fields.user.cost_center_id as user_cc,
    payload.fields.user.cost_center_name as user_cc_name,
    payload.fields.target.email as target_email,
    payload.fields.target.buy_org_id as target_buy_org,
    payload.fields.target.buy_org_name as target_buy_org_name,
    payload.fields.target.cost_center_id as target_cc,
    payload.fields.target.cost_center_name as target_cc_name

FROM "vtex"."schemaless"."b2b_suite_buyerorg_data_raw"
WHERE payload.name = 'b2b-suite-buyerorg-data'
and payload.kind in ('impersonate-ui-event', 'stop-impersonate-ui-event')
order by ingestion_time desc
```

![image](https://github.com/vtex-apps/b2b-organizations/assets/1576737/621db54c-cb9f-4c9e-87b7-a2d77ea0b10c)

#### How does this PR make you feel?

![](https://media.giphy.com/media/sTczweWUTxLqg/giphy.gif)
